mobile website

jo2 opened the pull request:
Issue: #7083 

This PR contains refactoring of the whole page except for project info and project documentation. Those parts are generated by `maven-project-info-reports-plugin`. For making these responsive the generation of those pages has to be changed. For example the `plugin-management.html` is generated in `PluginManagementReport.java`. The line which would have to be changed is [this](https://github.com/apache/maven-project-info-reports-plugin/blob/a06436366670e7fca7e92b690b2d45d5ada00ebc/src/main/java/org/apache/maven/report/projectinfo/PluginManagementReport.java#L202).

To make tables, images and code blockwise scrollable, those tags have to be surrounded by a `span` or `div` with the class `wrapper`. For example:
```
<img alt="example" src="example/path/image.png"/>
```
becomes
```
<div class="wrapper">
    <img alt="example" src="example/path/image.png"/>
</div>
```
The use of `div` and `span` depends on the block element which contains the image, code or table. In a `p` `span` has to be used, otherwise `div` can be used.

romani commented:
Thanks a lot for update.
 
We will need to ask you to fix indentation in xml ( as second commit of this PR), BUT only after we approve your PR.

Please deploy current version of site to your githubio https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/README.md#deploying-report , to let us review it on phones and on the go ( my laptop review time already have long queue).

I completely ok to keep non xdoc content as us for now, it rarely used.

jo2 commented:
When running diff.groovy from checkstyle tester on checkstyle itself I'm getting [this](https://jo2.github.io/) output. Doesn't seem right to me because all the docs are not included.. 

I used this command `groovy diff.groovy -r ..\..\checkstyle -b master -p #7083-mobile-website -c my_check.xml -l projects-to-test-on.properties -s true`. All used paths are valid so that is not the problem. In `checkstyle-tester/projects-to-test-on.properties` I removed guave and added checkstyle (I used the checkstyle entry which was in the comments line 7). 

Or did you mean I should publish the docs by using `mvn site:deploy`?

rnveach commented:
@jo2 `diff.groovy` is our regression utility for checks. It doesn't generate the xdoc website.
To generate the xdoc website use `mvn site`. The site is generated inside the `target` folder.

romani commented:
`mvn site -Pno-validations`

jo2 commented:
uploaded it to [https://jo2.github.io/](https://jo2.github.io/). Thanks!

romani commented:
![Screenshot_20191116-114406](https://user-images.githubusercontent.com/812984/68998436-35790a00-0867-11ea-9d65-cc34d82cf709.png)

Release dates still have problems, on mobile phone Android.

But in general update is awesome, all other pages rendering good.

jo2 commented:
Release Date is in PR https://github.com/checkstyle/checkstyle/pull/7254, but if it is easier i can merge those branches to one PR

romani commented:
Sorry my bad, so  I love this PR.

@rnveach , @pbludov , @strkkk , please review

rnveach commented on src/site/resources/css/site.css:
```diff
@@ -71,3 +71,31 @@ h2:hover .anchor, h3:hover .anchor {
   right: 0;
 }
+
+@media screen and (max-width: 767px) {
```
How was the 767 determined?

Desktop looks fine in full screen, but as I was resizing the window to make it smaller I noticed the left jumped from a small snippet of the left side to taking up the full window at about half screen. I was expecting something like this to kick in when the window got smaller than half, but I have never built screens for mobile before.

jo2 replied:
I can make i lower. 769px is the width of a tablet device. In my opinion, the normal design works fine on tablets. But yes, i think it would be better to choose a smaller value there. something like 600px should work better.

jo2 replied:
Due to the width of some of the table you'll get a horizontal scrollbar on some pages. But that problem occurs for 600px and 767px. Therefor I added the wrapper to the normal style. New version is up now.

strkkk commented:
@romani to he honest, I don't have enough knowledge in html/css to provide a good review for code.
New looks good on my phone (Android) and it is way better than current mobile version.

romani commented:
Any feedback is good, thanks a lot. At least you do not see visual problems and you like such change.

romani commented:
I played with this change on my local, looks good in FF on fullscreen and some sort of small size. Users usually use it one or another screen size. 
we should not care about  rendering in active resizing or please share video of that you see if that you think is critical

rnveach commented on src/xdocs/anttask.xml.vm:
```diff
@@ -91,6 +91,7 @@
     <section name="Parameters">
 
+      <div class="wrapper">
       <table summary="parameters">
```
would it be easier to always make the table itself a wrapper if it is part of the a section?

romani commented:
Looks like nobody has objection on change , that is good.
Now we need to satisfy CI to allow merge.
CI is failing on UT , that keep in sync javadoc and xdoc, as eventually we would like to have generate xdoc from javadoc.

There are 2 items we need to do:
- fix indentation in xml files if there is no way to reuse existing tags
- we need to sync extra xdoc modification to javadoc.

Extra tags will result in massive update, I am no afraid of it, but it is a lot of work.
Before we go the complicated way, is there any alternative ?

jo2 commented:
In my local branch I already have a fix for that but I ran into another issue with the JavaDoc not matching the expected value. I couldn't find where the expected value is defined though..

romani commented:
@jo2 , please be welcome to share problems, we will guide.

jo2 commented:
In `com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest` I get a ComparisonFailure because `AtclauseOrderCheck` `s class-leve JavaDoc was 
```
...7-02-28.
</p>
<ul>
<[li>
Property {@code violateExecutionOnNonTightHtml} - If turned on, will print violations if the Javadoc being examined by this check violates the tight html rules defined at <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML
Rules</a>. Default value is {@code false}.
</li>
<li>
Property {@code target} - Specify the list of targets to check at-clauses. Default value is <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CLASS_DEF">CLASS_DEF</a>, <a href="https://checkstyle.org/apidocs/com/pupp
ycrawl/tools/checkstyle/api/TokenTypes.html#INTERFACE_DEF">INTERFACE_DEF</a>, <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ENUM_DEF">ENUM_DEF</a>, <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/che
ckstyle/api/TokenTypes.html#METHOD_DEF">METHOD_DEF</a>, <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF">CTOR_DEF</a>, <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes
.html#VARIABLE_DEF">VARIABLE_DEF</a>.
</li>
<li>
Property {@code tagOrder} - Specify the order by tags. Default value is {@code @author, @deprecated, @exception, @param, @return, @see, @serial, @serialData, @serialField, @since, @throws, @version}.
</li>
<]/ul>
<p>
Default con...
```
instead of the expected
```
...7-02-28.
</p>
<ul>
<[]/ul>
<p>
Default con...
```

rnveach commented:
> `<[]/ul>`

Please look at https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java#L182-L188  and its surrounding code.
When we process XDoc, we expect everything to be at an exact level. Your changes made everything 1 level deeper, so the code needs to be updated to go 1 level deeper because of the `<div class="wrapper">`.

> In my local branch I already have a fix for that but I ran into another issue

I recommend pushing the changes you have, so we can duplicate your next problem and provide finer details.

jo2 commented:
Thanks, I'll have a look at it.

In regard to @rnveach`s [comment](https://github.com/checkstyle/checkstyle/pull/7254#pullrequestreview-318030034): In Firefox`s minimal width the release date and the release title are overlapping (s. Image). The minimal width is 286px though, problems arise below 320px for two digit version numbers (major and minor version). According to [this source](https://mediag.com/blog/popular-screen-resolutions-designing-for-all/) no problem should arise though.

![image](https://user-images.githubusercontent.com/7516893/69145781-bbe35500-0ace-11ea-96d1-ec032a83451e.png)

jo2 commented:
I can fix it for `Properties` subsections by skipping the `<div class="wrapper">` in the `populateProperties()` method but i don't think that is a appropriate way to handle it because there may be tables in every subsection. 

Is there a way to say that every node with the attribute `class` and the value `wrapper` should be skipped? Something like
```
if ("wrapper".equals(node.getAttributes.getNamedItem("class"))) {
    node = XmlUtil.getFirstChildElement(node);
}
```
This should be added in the tests though because the `Checker` class which causes the exception is a part of the checkstyle API which is used in other projects.

romani commented:
> Firefox's minimal width the release date and the release title are overlapping

It is, ideally should be no overlap, but it is very minor issue in comparison to general mobile support. Could be done later on in another PR.

>Something like

In such test all we need is to validate content between javadoc and xdoc, we can skip wrapping tags as much is reasonable. @pbludov, what is your opinion on this ?

jo2 commented:
@romani I fixed the problem regarding the ignoration of wrapper elements. Also, I pushed the identation fixes. @pbludov that is also fixed. please review. If there is anything to change, please comment so.

romani commented:
@rnveach , please review - https://romani.github.io/7083-mobile-website/index.html
I squashed commits to be more compact.

rnveach commented on src/xdocs/checks.xml:
```diff
@@ -898,14 +910,17 @@
-        </td>
-      </tr>
-    </table>
+      <div class="wrapper">
```
same.

jo2 replied:
the div has one more indent here because the table had the same indent as the wrapping section.

rnveach commented on src/xdocs/config_annotation.xml:
```diff
@@ -57,85 +57,87 @@ public String getNameIfPresent() { ... }
-          </tr>
-        </table>
+        <div class="wrapper">
+          <table>
```
How come this file indented table for the new div but `config.xml` didn't?

jo2 replied:
here the table had the correct indent regarding the wrapping elements

rnveach commented on src/xdocs/config.xml:
```diff
@@ -293,6 +293,7 @@
 
       <subsection name="Properties" id="Checker_Properties">
+        <div class="wrapper">
         <table>
```
See for previous comment.

jo2 replied:
answering to all conversations: some files had indents with a width of 2 spaces and some had for spaces. I can fix it if it is an issue.

rnveach replied:
It is best to reply to each individual one to prevent confusion.

For here, why wasn't `table` indented at all?

jo2 replied:
I seem to have missed this table. the problem with 4 and 2 spaces indents soesn't seem to be the reason for any of yout mentioned problems, sorry for wrong assumption.

jo2 commented:
@romani the difference in html file is because the checkstyle check for html files didn't pass due to the line being longer than 100 characters. Sometimes it failed for a 90 character line though. But I didn't find any regularity in it so no bug report til now.

@rnveach I can update the test adding a wrapper check for all tables, images, <pre>, and svg

rnveach commented:
> I can update the test adding a wrapper check for all tables, images, pre, and svg

Since you are checking all of those tags, it seems to me like this should be a new test and not the existing one.

> pre, and svg

I wasn't aware of these 2.
@romani you are ok with adding these wrappers everywhere just to be on the safe side?

jo2 commented on src/xdocs/checks.xml:
```diff
@@ -31,865 +31,877 @@
-        Outputs a JavaDoc tag as information.</td>
-      </tr>
-    </table>
+      <div class="wrapper">
```
the div has one more indent here because the table had the same indent as the wrapping section.

jo2 commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:
```diff
@@ -608,10 +608,15 @@ private static void validatePropertySection(String fileName, String sectionName,
             Assert.assertEquals(fileName + " section '" + sectionName
                     + "' subsection 'Properties' has unexpected child node",
-                "table", table.getNodeName());
+                "div", div.getNodeName());
```
check for `class=wrapper` is no problem. Should accept classList containing wrapper, too, though.

jo2 commented:
@rnveach pre is everything regarding code, source tags are rendered to a pre blog wrapped in a similar div like i did in this pr. svg is needed for the "become a backer" button in /sponsoring.html because that has a fix width of 600px. A solution here would be to limit the space the svg can take because its data has a fix width of 152px

romani commented:
>difference in html file is because the checkstyle check for html files didn't pass due to the line being longer than 100 characters.

I know, no complains.

>ok with adding these wrappers everywhere just to be on the safe side?

Sure, but let's do this in separate PR to not overcomplicate such good progress.

jo2 commented:
@romani In the commit 'add check for wrapper div' I already implemented the check and for the pipelines to work I had to add the wrappers everywhere (weren't that many locations anymore due to most wrappers alredy being set due to mobile website fix, though). 

Further I don't have insight in TeamCity and wrecker pipelines results. Why do they keep failing? something in my code? or is there already a fix like last time which I only have to merge?

romani commented:
wercker failure is unrelated ... you can ignore it for now

in teamcity you can login as guest and see all results
https://teamcity.jetbrains.com/viewLog.html?buildId=2652724&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest

please confirm that you can do this. Please fix issues.

meanwhile here is copy of violations

Selected: 'expression.equals("literal")' rather than '"literal".equals(expression)' (2)
src/test/java/com/puppycrawl/tools/checkstyle/internal
XdocsMobileWrapperTest.java (2)
84: iterateNode() Literal "div" is argument of 'equals()', instead of its qualifier 
85: iterateNode() Literal "span" is argument of 'equals()', instead of its qualifier 

Selected: Method may be 'static' (1)
src/test/java/com/puppycrawl/tools/checkstyle/internal
XdocsMobileWrapperTest.java (1)
78: iterateNode() Method iterateNode() may be 'static'

jo2 commented:
Thanks. Fix is pushed.

romani commented:
@rnveach , please finish review, release will in 1 day, it would good to merge this PR sooner.

romani commented:
rebased, squashed extra commits, did extra changes for proper indentation that I noticed.
Our original indentation was damaged in many places, this PR make it significantly better,  but there definitely a lot of room to improve (not in this PR).
I did conversion of google style and sun style xml file to 2 as indentation by IDEA.

pushed all changes to author fork.

romani commented:
non blocking regressions (can be fixed in separate PRs):
extra line wrapping after small images 
![image](https://user-images.githubusercontent.com/812984/69891951-5143cd80-12b5-11ea-9b74-20278b7c90d3.png)

extra space (if we try to make rendering awesome, but really ... not a problem):
![image](https://user-images.githubusercontent.com/812984/69891985-aaabfc80-12b5-11ea-96e8-83bfe4b9ffe7.png)
